pkg/rsnotify/listener/ipreporter.go

```diff
@@ -0,0 +1,62 @@
+package listener
+
+// Copyright (C) 2022 by RStudio, PBC.
+
+import "net"
+
+type IPReporter interface {
+	IP() string
+}
+
+type IPCache struct {
+	reporter    IPReporter
+	cachedValue string
+}
+
+// Note that this also implements the `IPReporter` interface, so there's really no need
+// to have a separate interface for the cache.
+func (c *IPCache) IP() string {
+	gotIP := c.reporter.IP()
+	if gotIP != c.cachedValue {
+		// enumerate IPs and determine if cached value is still value
+		if !c.stillValidIP(c.cachedValue) {
+			c.cachedValue = gotIP
+		}
+	}
+
+	return c.cachedValue
+}
+
+func (p *IPCache) stillValidIP(IP string) bool {
+	var ip net.IP
+	if ip = net.ParseIP(IP); ip == nil {
+		return false
+	}
+
+	interfaces, err := net.Interfaces()
```

Comment:
Instead of enumerating interfaces here and then following by enumerating the `inter.Addrs()`, I think we can use a single iteration of `net.InterfaceAddrs()` that gives us all the IP addresses at once.